### PR TITLE
Introduce and port over dependency injection articles for Microsoft Extensions

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

- Port over and adopt "Dependency Injection in .NET" article from the ASP.NET Core docs repo
- Split into three articles:
  - Overview
  - Tutorial
  - Guidelines
- Update TOC to link to our DI article
- Add supporting code samples

Contributes to: #4567

## Preview

✔️ [Dependency injection in .NET](https://review.docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection?branch=pr-en-us-20758)
✔️ [Tutorial: Use dependency injection in .NET](https://review.docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-usage?branch=pr-en-us-20758)
✔️ [Dependency injection guidelines](https://review.docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines?branch=pr-en-us-20758)

### CODE REVIEW

File: docs/core/extensions/dependency-injection-usage.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+> - Use service lifetime and scoping for DI
+
+## Prerequisites
+
+This tutorial uses:
+
+- [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet-core) or a later version.
+- An Integrated Developer Environment (IDE), [Visual Studio, Visual Studio Code, or Visual Studio for Mac](https://visualstudio.microsoft.com) are all valid choices
+- Ability to create new .NET applications
+- Familiarity with NuGet packages
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-usage.md:30]:
```suggestion
## Prerequisites

- [.NET Core 3.1 SDK](https://dotnet.microsoft.com/download/dotnet-core) or a later version.
- An Integrated Development Environment (IDE), [Visual Studio, Visual Studio Code, or Visual Studio for Mac](https://visualstudio.microsoft.com) are all valid choices.
- Familiarity with creating new .NET applications and installing NuGet packages.
```

File: docs/core/extensions/dependency-injection-usage.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+## Create a new console application
+
+Using either the [dotnet new](../tools/dotnet-new.md) command or the available IDE new project wizard, create a new .NET console application named **ConsoleDI.Example**. Add the [Microsoft.Extensions.Hosting](https://www.nuget.org/packages/Microsoft.Extensions.Hosting) NuGet package to the project.
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-usage.md:34]:
```suggestion
Using either the [dotnet new](../tools/dotnet-new.md) command or an IDE new project wizard, create a new .NET console application named **ConsoleDI.Example**. Add the [Microsoft.Extensions.Hosting](https://www.nuget.org/packages/Microsoft.Extensions.Hosting) NuGet package to the project.
```

File: docs/core/extensions/dependency-injection-usage.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+Finally, update the *Program.cs* file to match following:
+
+:::code language="csharp" source="snippets/configuration/console-di/Program.cs" range="1-18,35-60" highlight="22-26":::
+
+The application starts by:
+
+- Creating an <xref:Microsoft.Extensions.Hosting.IHostBuilder> instance with the [default binder settings](generic-host.md#default-builder-settings).
+- Services are configured and added with their corresponding service lifetime.
+- The host builder instance calls <xref:Microsoft.Extensions.Hosting.IHostBuilder.Build>, and assigns an instance of <xref:Microsoft.Extensions.Hosting.IHost>.
+- The `host` instance calls `ExemplifyScoping` passing in the <xref:Microsoft.Extensions.Hosting.IHost.Services?displayProperty=nameWithType>.
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-usage.md:91]:
```suggestion
The application:

- Creates an <xref:Microsoft.Extensions.Hosting.IHostBuilder> instance with the [default binder settings](generic-host.md#default-builder-settings).
- Configures services and adds them with their corresponding service lifetime.
- Calls <xref:Microsoft.Extensions.Hosting.IHostBuilder.Build> and assigns an instance of <xref:Microsoft.Extensions.Hosting.IHost>.
- Calls `ExemplifyScoping`, passing in the <xref:Microsoft.Extensions.Hosting.IHost.Services?displayProperty=nameWithType>.
```

File: docs/core/extensions/dependency-injection-usage.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+## Conclusion
+
+When running the application, you could expect to see similar output to the following:
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-usage.md:95]: The application produces output similar to the following example:

File: docs/core/extensions/dependency-injection-usage.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+```console
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-usage.md:97]:
```suggestion
```output
```

File: docs/core/extensions/dependency-injection-guidelines.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+description: Learn various dependency injection guidelines and best practices for .NET application development.
+author: IEvangelist
+ms.author: dapine
+ms.date: 09/23/2020
+ms.topic: guide
+---
+
+# Dependency injection guidelines
+
+In this article, you'll learn general guidelines and best practices for implementing dependency injection in your .NET applications. From design principles to proper disposal patterns, thread safety, and service lifetime management - these recommendations are tried and true.
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-guidelines.md:12]:
```suggestion
This article provides general guidelines and best practices for implementing dependency injection in .NET applications.
```

File: docs/core/extensions/dependency-injection-guidelines.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+## Design services for dependency injection
+
+When designing services for dependency injection:
+
+- Avoid stateful, static classes and members. Avoid creating global state by designing apps to use singleton services instead.
+- Avoid direct instantiation of dependent classes within services. Direct instantiation couples the code to a particular implementation.
+- Make services small, well-factored, and easily tested.
+
+If a class has many injected dependencies, it might be a sign that the class has too many responsibilities and violates the [Single Responsibility Principle (SRP)](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#single-responsibility). Attempt to refactor the class by moving some of its responsibilities into new classes. Keep in mind that Razor Pages page model classes and MVC controller classes should focus on UI concerns.
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-guidelines.md:22]: > Keep in mind that Razor Pages page model classes and MVC controller classes should focus on UI concerns.

Sounds like a comment for the ASP.NET Core edition of DI docs.

```suggestion
If a class has many injected dependencies, it might be a sign that the class has too many responsibilities and violates the [Single Responsibility Principle (SRP)](/dotnet/standard/modern-web-apps-azure-architecture/architectural-principles#single-responsibility). Attempt to refactor the class by moving some of its responsibilities into new classes.
```

File: docs/core/extensions/dependency-injection-guidelines.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+}
+```
+
+In the preceding code:
+
+- The `ExampleService` instance is **not** created by the service container.
+- The framework does **not** dispose of the services automatically.
+- The developer is responsible for disposing the services.
+
+### IDisposable guidance for Transient and shared instances
```
Comment (tdykstra) [inline docs/core/extensions/dependency-injection-guidelines.md:79]: The text in these sections is hard to follow without sample code.

**Review by tdykstra — APPROVED** [review]
Great content, clear and concise -- a few optional suggestions for improvement.

File: docs/core/extensions/dependency-injection-usage.md (hunk 5 above)
Comment (IEvangelist) [reply docs/core/extensions/dependency-injection-usage.md:97]: How does this render?
